### PR TITLE
switch to static builds

### PR DESCRIPTION
**Pull request by jgraettinger**
Use (a fork of) rust-rocksdb to actually build RocksDB.
A librocks-exp crate is then added to make the RocksDB artifacts
available at a stable workspace location.

Update build infra to use a new go.sh script, which is now the one and
only place where various CGO flags are set up to find and link
appropriate libraries. go.sh is a very minimal wrapper around the `go`
command to also support development workflows.

Update Rust dependencies, and fix usages of sqlite and rust-rocksdb
accordingly. Notably, rocksdb::Env is no longer a thin wrapper around
the FFI environment, and transumation is now very broken. Our fork has a
new constructor to make this safer and stable.

---
This change is [Reviewable](https://reviewable.io/reviews/estuary/flow/172)

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,4 +1,5 @@
 [workspace]
+resolver = "2"
```
Comment (saterus) [inline Cargo.toml:2]: Did this end up fixing things for you? 

Last night you said it would spit out warnings about duplicated dependencies, but when I compiled, I didn't see anything like that. Expected?

File: Makefile [hunk 2]
```diff
@@ -68,33 +60,28 @@ PACKAGE_TARGETS = \
...
 
 ##########################################################################
 # Build rules:
 
 # `protoc-gen-gogo` is used to compile Go protobufs.
 ${TOOLBIN}/protoc-gen-gogo:
-	go mod download github.com/golang/protobuf
-	go build -o $@ github.com/gogo/protobuf/protoc-gen-gogo
+	./go.sh mod download github.com/golang/protobuf
+	./go.sh build -o $@ github.com/gogo/protobuf/protoc-gen-gogo
```
Comment (saterus) [inline Makefile:84]: Not sure if this is a problem, or just something to note:

I had to remove my `.build/tools` directory and rerun `make install-tools` to get `make rust-build` to work. Prior to that, it was failing to find the compiled protobuf stuff. :thinking:

File: crates/derive/Cargo.toml [hunk 3]
```diff
@@ -15,9 +15,10 @@ anyhow = "*"
 bytes = "*"
 futures = "*"
 itertools = "*"
+librocksdb-sys = { version = "*", features = ["lz4", "rtti"] }
 pin-utils = "*"
 prost = "*"
-rocksdb = { version = "*", default-features = false }
+rocksdb = { version = "*", default-features = false, features = ["lz4", "rtti"] }
```
Comment (saterus) [inline crates/derive/Cargo.toml:21]: Do we need to specify the features on both of these libraries explicitly? Would specifying them on `rocksdb` alone apply them to `librocksdb-sys` as well?

File: crates/models/src/tables/macros.rs [hunk 4]
```diff
@@ -316,7 +316,7 @@ macro_rules! tables {
 
             fn load_all(&mut self, db: &rusqlite::Connection) -> rusqlite::Result<()> {
                 let mut stmt = db.prepare(&Self::select_sql())?;
-                self.0.extend(stmt.query_map(rusqlite::NO_PARAMS, $row::scan)?
+                self.0.extend(stmt.query_map([], $row::scan)?
```
Comment (saterus) [inline crates/models/src/tables/macros.rs:319]: Are these changes related to the static build?

**Review by saterus — APPROVED** [review]
I've tested it out and it seems to work on my machine. :tada:

No blockers, just curious about a few things.

**Review by jgraettinger — COMMENTED** [review]
Thanks!

> *[Reviewable](https://reviewable.io/reviews/estuary/flow/172#-:-Mfixc3592aqrtt833wT:b-ucfknh)* status: 0 of 18 files reviewed, 4 unresolved discussions (waiting on @psFried and @saterus)

___
*[Cargo.toml, line 2 at r1](https://reviewable.io/reviews/estuary/flow/172#-Mfdbroiv2XXIctqZzzi-r1-2:-MfiwpQw527059OTWguu:bmzieme) ([raw file](https://github.com/estuary/flow/blob/1dd2c52b1aa0f597846f87b9a68cf4ae015e7517/Cargo.toml#L2)):*
<details><summary><i>Previously, saterus (Alex Burkhart) wrote…</i></summary><blockquote>

Did this end up fixing things for you? 

Last night you said it would spit out warnings about duplicated dependencies, but when I compiled, I didn't see anything like that. Expected?
</blockquote></details>

It didn't -- it's still building the librocksdb-sys crate twice, and I have no idea why.

I've been through the various tips in the feature unification docs, and AFAIK they're being built using identical feature flags.

Basically I kicked the can on this, and am hoping that existing caching of the `target` directory will do the trick.
___
*[Makefile, line 84 at r1](https://reviewable.io/reviews/estuary/flow/172#-Mfdbroiv2XXIctqZzzj-r1-84:-MfixHWUCnxDTf45jxmv:bqv2ty0) ([raw file](https://github.com/estuary/flow/blob/1dd2c52b1aa0f597846f87b9a68cf4ae015e7517/Makefile#L84)):*
<details><summary><i>Previously, saterus (Alex Burkhart) wrote…</i></summary><blockquote>


Not sure if this is a problem, or just something to note:

I had to remove my `.build/tools` directory and rerun `make install-tools` to get `make rust-build` to work. Prior to that, it was failing to find the compiled protobuf stuff. :thinking: 
</blockquote></details>

Huh, got me.
___
*[crates/derive/Cargo.toml, line 21 at r1](https://reviewable.io/reviews/estuary/flow/172#-Mfdbroiv2XXIctqZzzk-r1-21:-MfixNhr7tYoAMxfi_5m:b-h21pl6) ([raw file](https://github.com/estuary/flow/blob/1dd2c52b1aa0f597846f87b9a68cf4ae015e7517/crates/derive/Cargo.toml#L21)):*
<details><summary><i>Previously, saterus (Alex Burkhart) wrote…</i></summary><blockquote>

Do we need to specify the features on both of these libraries explicitly? Would specifying them on `rocksdb` alone apply them to `librocksdb-sys` as well?
</blockquote></details>

I didn't try. My understanding w/ the v2 resolver is that it _would_ consider them separately built targets, but I didn't test.
___
*[crates/models/src/tables/macros.rs, line 319 at r1](https://reviewable.io/reviews/estuary/flow/172#-Mfdbroiv2XXIctqZzzm-r1-319:-MfixXsBBrO9eVIGjXmi:b-gv9ygu) ([raw file](https://github.com/estuary/flow/blob/1dd2c52b1aa0f597846f87b9a68cf4ae015e7517/crates/models/src/tables/macros.rs#L319)):*
<details><summary><i>Previously, saterus (Alex Burkhart) wrote…</i></summary><blockquote>

Are these changes related to the static build?
</blockquote></details>

They're fixing deprecation warnings that came up in a newer version of the rusqlite crate.

**Review by psFried — CHANGES_REQUESTED** [review]
Wahoo! I'm excited for this to land. I left a few minor comments, but otherwise looks great.

> Reviewed 18 of 18 files at r1.
> *[Reviewable](https://reviewable.io/reviews/estuary/flow/172#-:-MfiytAnClbp-Jj2Dgko:b-wawt4i)* status: all files reviewed, 6 unresolved discussions (waiting on @jgraettinger and @saterus)

___
*[Makefile, line 123 at r1](https://reviewable.io/reviews/estuary/flow/172#-MfivBmkBoxqr-1_xnT7:-MfivBmkBoxqr-1_xnT8:bh0jinl) ([raw file](https://github.com/estuary/flow/blob/1dd2c52b1aa0f597846f87b9a68cf4ae015e7517/Makefile#L123)):*
> ```Makefile
> 	FLOW_VERSION=${VERSION} cargo build --release --locked -p bindings
> 
> ${RUSTBIN}/librocks-exp/librocksdb.a:
> ```

As written, this task won't get re-run if you update the code in librocks-exp. I think the best way to handle this would be to add `$(shell find librocks-exp -type f)` as a prerequisite.
___
*[crates/librocks-exp/src/lib.rs, line 0 at r1](https://reviewable.io/reviews/estuary/flow/172#-MfixzEUBTW6BGIeagiN:-MfixzEV51gWMFkNhc1e:b-1aqpdf) ([raw file](https://github.com/estuary/flow/blob/1dd2c52b1aa0f597846f87b9a68cf4ae015e7517/crates/librocks-exp/src/lib.rs#L0)):*
I'm thinking it'll be helpful to future folk to have a README, or maybe just a comment at the top of this file that explains why this crate is empty. Disregard if you already have one and I just haven't seen it yet.

**Review by jgraettinger — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/estuary/flow/172)* status: all files reviewed, 6 unresolved discussions (waiting on @jgraettinger, @psFried, and @saterus)

___
*[Makefile, line 123 at r1](https://reviewable.io/reviews/estuary/flow/172#-MfivBmkBoxqr-1_xnT7:-Mfj1Uyj5S3IOD3jW6xm:b57efpi) ([raw file](https://github.com/estuary/flow/blob/1dd2c52b1aa0f597846f87b9a68cf4ae015e7517/Makefile#L123)):*
<details><summary><i>Previously, psFried (Phil Fried) wrote…</i></summary><blockquote>

As written, this task won't get re-run if you update the code in librocks-exp. I think the best way to handle this would be to add `$(shell find librocks-exp -type f)` as a prerequisite.
</blockquote></details>

Hmm I'm also now worried about a cached `target` directory from a non-matched Cargo.lock being used, and this silently failing to run...

**Review by jgraettinger — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/estuary/flow/172)* status: all files reviewed, 6 unresolved discussions (waiting on @jgraettinger, @psFried, and @saterus)

___
*[Makefile, line 123 at r1](https://reviewable.io/reviews/estuary/flow/172#-MfivBmkBoxqr-1_xnT7:-Mfj4K8vDfwDf2GgB9-M:bt5sqn7) ([raw file](https://github.com/estuary/flow/blob/1dd2c52b1aa0f597846f87b9a68cf4ae015e7517/Makefile#L123)):*
<details><summary><i>Previously, jgraettinger (Johnny Graettinger) wrote…</i></summary><blockquote>

Hmm I'm also now worried about a cached `target` directory from a non-matched Cargo.lock being used, and this silently failing to run...
</blockquote></details>

Right we do currently have this "files could have changed" problem with `libbindings.a` also. For me it's meant "run a manual `make rust-build` before building `flowctl`" during development.

It would also be an issue w/ CI, except that we don't use `restore-keys` anymore and require an exact match on the Cargo.lock hash.

**Review by jgraettinger — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/estuary/flow/172)* status: 17 of 18 files reviewed, 6 unresolved discussions (waiting on @psFried and @saterus)

___
*[crates/librocks-exp/src/lib.rs, line  at r1](https://reviewable.io/reviews/estuary/flow/172#-MfixzEUBTW6BGIeagiN:-Mfj729h7U_28Il3yMcv:b-896fix) ([raw file](https://github.com/estuary/flow/blob/1dd2c52b1aa0f597846f87b9a68cf4ae015e7517/crates/librocks-exp/src/lib.rs#L)):*
<details><summary><i>Previously, psFried (Phil Fried) wrote…</i></summary><blockquote>

I'm thinking it'll be helpful to future folk to have a README, or maybe just a comment at the top of this file that explains why this crate is empty. Disregard if you already have one and I just haven't seen it yet.
</blockquote></details>

Done.
